[Roach 미션2: 체스판 생성 구현]

tmdgusya opened the pull request:
Dion 의 피드백을 받고 많은 코드 수정이 있었다.
하지만 오늘자 호눅스의 수업을 듣고 미션 요구사항에만 그냥 집중하기로 했다.
그래서 Pointer 같은 먼 미래에 필요할 수도 있는 것들은 없애 버리고, 가까이에 있는 것들만 적었습니다.
따라서 현재 존재하는 클래스는 Board, Pawn, Color, Piece 입니다.

매개 변수 없는 생성자
이 테스트는 왜 컴파일 에러를 발생시키는가? 

제 코드를 기준으로는 Color 에 final 이 지정되어 있기에 첫 생성시 Color 를 주입하지 않으면 컴파일 타입 에러가 나게 됩니다.
그래서 아래와 같은 코드로 변경하였고, 컴파일 타입 에러를 고칠수 있었습니다.
```java
    public Pawn(){
        super('p', Color.WHITE);
    }
```

해당 Color 부분은 대입하지 

## 추가 테스트
```java
   @Test
    @DisplayName("Piece 객체 이외의 객체를 등록하면 컴파일 에러가 난다.")
    void addOtherObjectReturnException(){
        board.add(new Integer("7"));
    }
```
> result
```
* What went wrong:
Execution failed for task ':compileTestJava'.
> Compilation failed; see the compiler error output for details.
```
컴파일 에러가 나는것을 확인하였고, 그런데 빨간줄이 계속 남아있어 테스트 코드에서는 일단 지웠습니다.
컴파일 오류가 뜨는 이유는 board.add(Pawn pawn) 으로 이루어져 있는데, 알맞지 않은 타입인 Integer 가 provide 되었기 때문에
컴파일 시점에서 이를 오류로 인지하고 컴파일 오류를 발생시키기 때문입니다.

해당 코드를 통과시키기 위해서는 board.add(Object o) 로 바꿀 수 있으나, 이렇게 되면 board.add 하는 것들은 Pawn 이여야 하는데
목적이 불분명해져 안좋은 코드, 혹은 목적이 불분명한 메소드가 된다고 생각합니다.

* 요구사항에 대해 대폭 수정하면서 커밋이 불필요하게 많아진것 같다는 생각이 듭니다.
* 앞으로는 커밋시 주의해야겠다는 생각이 들게 되었습니다.

tmdgusya committed:
[1차 코드리뷰 사항 반영]
- build.gradle
  - 사용하지 않는 Junit 구버전 의존성 코드 제거
- Piece.java
  - 객체 자체의 책임 및 2차 요구사항을 살펴보니 name 필드 필요하다고 생각하여 추가
  - 코드 형태 reformatting 실시
  - setCurrentPointer() 객체 자체의 값을 받아와서 활용하는 방안으로 변경
  - kill() -> killPiece() 로 메소드가 하는일을 객체지향적으로 좀더 명확하게 이름 변경
  - 초기에 입력되어야 하는 Pointer 도 final 로 변
- Pawn.java
  - 부모 클래스 변동으로 인한 생성자 변경
- Pointer.java
  - 편의 메소드를 삭제하고, 객체 자체를 활용할 수 있는 로직으로 변경함.
- PawnTest.java
  - Pawn 생성시 Constructor 에 생성되는 값들에 대한 유효성 검사 테스트 케이스 추가

tmdgusya committed:
[2차 요구사항 반영]

Board.java
 - board 라는 8*8 판을 2차원 배열을 사용하여, 체스칸과 동일한 크기를 지닐 수 있도록 함.
 - pawnList 요구사항에서 주어진 pawn 을 ArrayList 로 관리할 수 있도록 함
   - Generic 을 통해서 Pawn 에 타입 제한을 걸어두어서, 컴파일 시 다른 객체를 넣으면 오류남.
 - add(Pawn pawn) 으로 Pawn 객체만 추가하여 넣을 수 있도록 함.
 - size() pawn 이 얼마나 있는지 size() 를 리턴해주도록함
 - findPawn(index) index 를 받아와서 해당 index 의 pawn 을 리턴해준다.

BoardTest.java
 - create()
   - 요구사항에 나와있는 테스트를 그대로 구현하였음
 - exceptionTest()
   - 폰이 없는데도 불구하고 가져오려할때 IndexOutOfBoundsException 이 리턴되는지 확인.

Piece.java
 - 기존에는 색깔을 구분하기위해 두글자씩 하다보니 String 으로 유지했으나, 요구사항 대로면
   한글자로 진행될것 같아. 메모리 공간의 절약을 위해 char 로 타입을 변경함.

tmdgusya committed:
type: CODE REVIEW 반영
Subject : Dion 이 남겨준 코드리뷰 답변들에 대한 수정사항 반영

1. Pointer.java
  - Pointer 객체 자체에서의 책임을 높이기위해 setPointer() 편의 메소드 작성
  - separator field 를 static 으로 작성
    - 이는 Class 당 하나만 생성되어도 충분한 변수라고 생각되기때문
2. Piece.java
  - currentPointer 및 alive 'final' 제거
    - 계속 변동될 수 있는 객체의 상태를 나타내는 필드들이기에 final 이 맞지 않는다고 생각됨
3. PawnTest.java
  - 디렉토리 변동으로 인한 package 명 수정

tmdgusya committed:
[Roach] 미션2 : 체스판 생성 구현 완료

type: 요구사항 구현
Subject : Mission2 체스판 만들기 요구사항 구현

1. Board.java
  - CHESS_BOARD_SIZE 를 8*8 인 체스판의 최대크기인 8인 상수로서 작성함
    - 후에 체스크기를 이용해야 한다면 이를 private -> public 으로 변경해도 된다고 생각됨.
  - board : Piece 객체들이 존재할 수 있는 배열로써 8*8 의 크기를 지님.
    - Piece 들의 배열로써 Piece 들이 어디에든 위치할 수 있음.
  - pawnList : pawn 을 관리하는 ArrayList
    - 후에 pawn 개수 불러오기 등을 사용할때 다 같이 관리하는것 보다는 말 단위로 관리하는 것이 빠     르게 탐색할 수 있다고 판단됨.
  - add() : pawnList 에 pawn 을 추가할 수 있는 기능
  - size() : 현재 pawn 이 몇개나 있는지 파악할 수 있는 메소드
  - findPawn() : 현재 Pawn 을 보관할 수 있는 List 에서 Pawn 을 꺼내올 수 있는 메소드
    - 존재하지 않는 인덱스에서 가져오려고 할시 IndexOutOfBoundsException 발생 해당 부분 예     외처리하여 런타임 시 종료되지 않도록 처리함.

2. BoardTest.java
  - beforeEach() : board 를 테스트 메소드 하나당 재 생성하게함. 테스트간 영향을 주지 않기 위해
  - create() : 생성한 Pawn 들을 보드에 등록하여 정상적으로 등록되고 관리되어 지고 있는지를 테스   트 함.
  - exceptionTest() : findPawn 메소드 이용간 정상적이지 않은 범위를 index 값으로 입력했   을때 정상적으로 IndexOutOfBoundsException 이 발생하는지 확인함.

tmdgusya committed:
type : fix

코드스쿼드에서 제공하는 요구사항을 준수하기 위해 대폭 변경
- Pointer.class 제거
- Pawn.class
  - 기본 생성자로 생성시 WHITE 컬러로 생성되도록 변경

tmdgusya committed:
type : test

Pawn TEST 내용 추가
- 기본 생성자로 생성할시 WHITE 인지 확인

tmdgusya committed:
type : fix

Color 가 주입되지 않았을때 컴파일 에러로 추가해주기 위해 Color 'final' 추가

tmdgusya committed:
type : fix

Color, name 이 주입되지 않았을때 컴파일 에러로 추가해주기 위해 Color name'final' 추가

ksundong commented on .idea/aws.xml:
```diff
@@ -0,0 +1,11 @@
+      </list>
+    </option>
+  </component>
+</project>
```
이 파일은 현재 미션하고는 관련이 없어보입니다.
협업을 하실 때에는 요런 파일은 `.gitignore`를 설정하셔서 git의 버전관리 대상이 되지 않도록 해주는 것도 좋습니다. (일단 진행해주세요!)

tmdgusya replied:
앗 제가 알고리즘 bash 랑 헷갈려서 실수로 git add . 을 친거같네요..
원래는 IDE툴을 이용해서 하는데 gitignore 를 설정해놔야겠어요 감사합니다

ksundong commented on .idea/codeStyles/Project.xml:
```diff
@@ -0,0 +1,7 @@
+<component name="ProjectCodeStyleConfiguration">
+  <code_scheme name="Project" version="173">
+    <ScalaCodeStyleSettings>
+      <option name="MULTILINE_STRING_CLOSING_QUOTES_ON_NEW_LINE" value="true" />
+    </ScalaCodeStyleSettings>
```
스칼라 코드 스타일도 들어왔군요 ㅋㅋㅋ

ksundong commented on build.gradle:
```diff
@@ -9,8 +9,7 @@ repositories {
     testCompile("org.assertj:assertj-core:3.18.1")
-}
\ No newline at end of file
+}
```
👍

ksundong commented on src/main/java/net/honux/chess/attribute/Pointer.java:
```diff
@@ -2,30 +2,38 @@
-        this.x = pointer1.split(separator)[0].charAt(0);
-        this.y = Integer.parseInt(pointer1.split(separator)[1]);
+    public void setPointer(Pointer pointer) {
+        this.x = pointer.getX();
+        this.y = pointer.getY();
```
```suggestion
        this.x = pointer.x;
        this.y = pointer.y;
```

이렇게도 가능합니다. 한 번 해보세요~
Java에서는 같은 타입의 속성을 참조할 수 있습니다.

tmdgusya replied:
아 그렇군요. 
이런 생각을 못해봤는데 감사해요 Dion ! :)

ksundong commented on src/main/java/net/honux/chess/attribute/Pointer.java:
```diff
@@ -2,30 +2,38 @@
+    public void setY(int y) {
+        this.y = y;
+    }
 }
```
리뷰를 하고 보니, 이 파일 없앴다고 하셨는데 그대로 남아있군요?

tmdgusya replied:
아 삭제하겠습니다. 
upstream/fetch 하면서 들어온것 같은데, 제가 뭔가 실수를 한것 같습니다.

ksundong commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -0,0 +1,31 @@
+public class Board {
+    private final static int CHESS_BOARD_SIZE = 8;
+    private final Piece[][] board = new Piece[CHESS_BOARD_SIZE][CHESS_BOARD_SIZE];
+    ArrayList<Pawn> pawnList = new ArrayList<>();
```
음 접근제어자를 놓치셨군요.
항상 신경써주셔야 합니다. ^^

ksundong replied:
그리고 컬렉션을 선언할 때, `List`로 선언해도 되는데요. 그 이유는 무엇일까요?
대부분의 Java 개발에서는 `List`로 선언하는 방식을 선호하는 것과 연관이 있습니다!

tmdgusya replied:
구현체로 타입을 작성하게 되면 후에 바꾸게 될때 다른 리스트를 쓰지 못하게 되기에
List 와 같은 추상타입을 타입으로 해놓으면, 어떤 List를 구현한 구현체를 넣어도 상관없는 것과, 객체 자체에서 구현체에 의존을 
하지 않게되기 때문인 것 같습니다. List 로 바꾸겠습니다.

감사해요 Dion 이런 디테일한 부분 놓치면 안되는데 놓치게되네요!
그래도 코드 리뷰덕분에 다시끔 상기시키게 되었습니다.

ksundong commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -0,0 +1,31 @@
+
+public class Board {
+    private final static int CHESS_BOARD_SIZE = 8;
+    private final Piece[][] board = new Piece[CHESS_BOARD_SIZE][CHESS_BOARD_SIZE];
```
요 `board`와 `pawnList`의 역할을 명확히 구분해주셔야 할 것 같아요.
`board`는 사용되지 않는 느낌입니다.

tmdgusya replied:
board 는 일단 없애야 할것 같습니다. :)

ksundong commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -0,0 +1,31 @@
+        }catch (IndexOutOfBoundsException e){
+            System.out.println("해당 위치에 Pawn 이 존재하지 않습니다.");
+        }
+        return pawn;
```
이 코드는 `null`을 리턴하게 될텐데 `null`을 리턴하는 것은 삼가도록 합시다.

이 단계에서 학습하시기 버거울 수도 있는데 `Optional`에 대해서 한 번 알아보세요.
어려우면 학습은 잠시 미뤄두셔도 됩니다.

tmdgusya replied:
감사합니다. 전해 학습했던 내용을 바탕으로 아래와 같은 코드로 변경하였습니다.
```java
  public Optional<Pawn> findPawn(int index) {
        Optional<Pawn> pawn = Optional.empty();
        try {
            pawn = Optional.ofNullable(pawnList.get(index));
        }catch (IndexOutOfBoundsException e){
            System.out.println("해당 위치에 Pawn 이 존재하지 않습니다.");
        }
        return pawn;
    }
```

ksundong commented on src/main/java/net/honux/chess/entity/pieces/Bishop.java:
```diff
@@ -0,0 +1,26 @@
+import net.honux.chess.attribute.Color;
+import net.honux.chess.attribute.Pointer;
+
+public class Bishop extends Piece {
```
현재 요구사항엔 없는 클래스가 추가되었군요?

tmdgusya replied:
이것또한 삭제하겠습니다.
git add . 때문에 여간 이상한 클래스들이 들어왔네요

ksundong commented on src/main/java/net/honux/chess/entity/pieces/Main.java:
```diff
@@ -0,0 +1,13 @@
+
+public class Main {
+
+    public static void main(String[] args) {
+        Board board = new Board();
+        board.findPawn(1);
+    }
```
이 main은 왜 작성해주셨는지 알 수 있을까요?

tmdgusya replied:
아 그 저번 테스트간 잠시 동작을 보려고 했는데 git add . 되면서 추가되버렸네요

ksundong commented on src/main/java/net/honux/chess/entity/pieces/Pawn.java:
```diff
@@ -1,20 +1,26 @@
 
-    public Pawn(Color color, Pointer startPointer) {
-        super(color, startPointer);
+//    public Pawn(){ }
```
쓸모 없어진 코드는 주석처리 하지 마시구 지워주시면 됩니다!

tmdgusya replied:
네 알겠습니다!

ksundong commented on src/main/java/net/honux/chess/entity/pieces/Piece.java:
```diff
@@ -1,31 +1,28 @@
-import net.honux.chess.attribute.Pointer;
 
 public abstract class Piece {
+    private final char name;
```
요거는 왜 필요한가요?

tmdgusya replied:
'p' 와 'P' 를 넣어주려고 했는데, 현재 구현단계에서는 필요없는것 같습니다.
제거하겠습니다. 사실 Mission3 으로가도, Color에 따라 구분되게 내부로직으로 처리하는게 더 맞는것 같기도합니다.

ksundong commented on src/main/java/net/honux/chess/entity/pieces/Piece.java:
```diff
@@ -1,31 +1,28 @@
+        return color;
     }
+
 }
```
사실 현재 Piece도 필요없는 클래스입니다.
Pawn만 가지고 문제를 충분히 해결하실 수 있어요.

tmdgusya replied:
Piece 도 삭제하도록 하겠습니다!

ksundong commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    @DisplayName("Pawn 이 보드에 정상적으로 등록되어야 한다.")
+    void create() {
```
메서드명을 읽는 사람이 알기 쉽도록 작성해주시는 것도 좋을 것 같아요.

tmdgusya replied:
넵 알겠습니다.
해당 메소드 네임도 가독성을 높여보도록 하겠습니다.

ksundong commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -0,0 +1,38 @@
+    void create() {
+        Pawn white = new Pawn('p', Color.WHITE);
+        board.add(white);
+        Assertions.assertEquals(1, board.size());
+        Assertions.assertEquals(white, board.findPawn(0));
```
AssertJ의 `assertThat()` 메서드를 활용해주세요.
또, 여러개의 Assertion을 사용하기 보다는 `assertAll()`을 활용해주세요.
이 [링크](https://stackoverflow.com/questions/40796756/assertall-vs-multiple-assertions-in-junit5)를 참고해주세요!

tmdgusya replied:
넵 알겠습니다.
감사합니다. 앞으로는 assertAll 로 적는 습관을 가져야 겠군요!

ksundong commented on src/test/java/net/honux/chess/entity/pieces/PawnTest.java:
```diff
@@ -0,0 +1,46 @@
+        assertThat(result).isEqualTo(expectedColor);
+    }
+
+    @Test
+    @DisplayName("폰의 이름이 정상적으로 등록되어야 한다.")
+    void setPawnName() {
+        Pawn whitePawn = new Pawn('p', Color.WHITE);
+        Pawn blackPawn = new Pawn('p', Color.BLACK);
+        assertThat(whitePawn.getName()).isEqualTo('p');
+        assertThat(blackPawn.getName()).isEqualTo('p');
+    }
```
다 똑같은 이름이라면 그냥 변수로 빼주는 편이 나을 것 같습니다.

ksundong replied:
그리고 굳이 테스트를 두 번 하지 않아도 되는 코드입니다.
둘의 관심사는 색상이 아니라 이름이기 때문이에요.

tmdgusya replied:
넵 알겠습니다. 일단 name field 가 없어지면서 삭제되긴 하였지만,
테스트 하는데 있어서도 관심사에 유의하겠습니다.

ksundong commented on src/test/java/net/honux/chess/entity/pieces/PawnTest.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+
+    @Test
+    @DisplayName("Pawn 을 기본 생성자로 생성할시 하얀색으로 생성되는지 확인한다.")
+    void PawnTest(){
+        Pawn pawn = new Pawn();
+        assertThat(Color.WHITE).isEqualTo(pawn.getColor());
+    }
```
👍

ksundong reviewed (changes requested):
안녕하세요. Roach!! 리뷰어 디온입니다. 두 번째 PR 까지 잘 오고 계십니다!
이번 PR은 작성해주신 PR 내용이 현재 푸시된 항목들과 일치하지 않아서 변경을 요청드립니다.

## 변경 요구 사항

- `Board`, `Pawn`, `Color`를 제외한 클래스를 제거시켜주세요. (테스트 코드만 남기구요.)
- 나중 커밋으로 갈 수록 커밋 제목이 이 커밋이 무엇을 하는지 잘 알 수 없습니다.
   ![image](https://user-images.githubusercontent.com/38597469/107597776-049ca600-6c5f-11eb-8775-c132415a4be4.png)
- 관련 없어 보이는 파일은 현재는 그냥 두고, 다음 미션을 하실 때, `gitignore`에 대해서 학습을 하고 적용시켜주세요.
- Board에서 `board`와 `pawnList`가 둘 다 있는 이유는 무엇인가요? `board`는 사용되지 않고 있습니다. 불필요한 코드는 제거시켜주세요.
- `null`을 리턴하는 코드는 가급적이면 삼가주시는게 좋습니다. `NullPointerException`이 발생할 여지를 주지 않도록 코딩을 해주세요!
- `assertThat()`을 사용해주시고 JUnit assertion은 사용하지 않는 것을 추천드려요.

이번 미션 요구사항은 충족하셨지만, 위의 사유들로 변경 요청드립니다!
고생많으셨습니다!

tmdgusya committed:
type : fix

필요없는 파일들 제거

tmdgusya committed:
type : review

코드리뷰 내용 반영

- Board.java
  - 필요없는 field board 제거
  - pawnList 의 타입은 구현체가 아닌 추상인터페이스 타입으로 교체
  - findPawn 메소드 Optional 을 반환하도록 변경

- BoardTest.java
  - create() 보다는 createPawnOntheBoard() 로 가독성 있게 변경
  - Assertions 중복되는 것들 assertAll() method 로 통일.

- Pawn.java
  - Piece 제거로 인해 Field 변수 추가

- PawnTest.java
  - name Field 제거로 인해 생성자 수정

tmdgusya commented:
@ksundong 
바꾼 내용으로 재 푸시 했습니다.
Commit 에 fix 와 review commit 으로 등록해 두었습니다.

Dion 감사해요!
명절 잘 보내세요~~!😁

ksundong commented on src/main/java/net/honux/chess/core/Board.java:
```diff
@@ -0,0 +1,30 @@
+        return pawnList.size();
+    }
+
+    public Optional<Pawn> findPawn(int index) {
+        Optional<Pawn> pawn = Optional.empty();
+        try {
+            pawn = Optional.ofNullable(pawnList.get(index));
+        }catch (IndexOutOfBoundsException e){
+            System.out.println("해당 위치에 Pawn 이 존재하지 않습니다.");
+        }
+        return pawn;
```
오 `Optional`을 공부해보셨군요. 좋습니다. 👍 
제가 생각했던 코드와 비슷해진 것 같아요.

ksundong commented on src/test/java/net/honux/chess/core/BoardTest.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Test
+    @DisplayName("Pawn 이 보드에 정상적으로 등록되어야 한다.")
+    void createPawnOntheBoard() {
+        Pawn white = new Pawn(Color.WHITE);
+        Pawn black = new Pawn(Color.BLACK);
+        board.add(white);
+        Assertions.assertAll(
+                ()->assertThat(board.size()).isEqualTo(1),
+                ()->assertThat(white).isEqualTo(board.findPawn(0).get()));
+        board.add(black);
+        Assertions.assertAll(
+                ()->assertThat(board.size()).isEqualTo(2),
+                ()->assertThat(black).isEqualTo(board.findPawn(1).get()));
```
`assertAll()` 적용 좋습니다.
하지만, 되도록 테스트는 테스트 하고자 하는 부분을 아래쪽으로 몰아주세요.
테스트 하고자 하는 바를 더 명확히 알 수 있도록 가독성이 좋아지게 됩니다.

저라면 아래와 같이 구성했을 것 같습니다.
```suggestion
    void createPawnOntheBoard() {
        Pawn white = new Pawn(Color.WHITE);
        Pawn black = new Pawn(Color.BLACK);
        
        board.add(white);
        board.add(black);
        
        assertAll(
                () -> assertThat(board.size()).isEqualTo(2),
                () -> assertThat(white).isEqualTo(board.findPawn(0).get()));
                () -> assertThat(black).isEqualTo(board.findPawn(1).get()));
```

tmdgusya replied:
그렇군요 AssertAll() 을 저렇게 생각하면 좋겠군요.

커밋 메세지 관련한건 이번 연휴간 읽어보고 커밋 템플릿을 설정해놓아야겠군요

ksundong replied:
ㅎㅎ 커밋메시지는 규칙성이 있는게 보기좋긴 합니다.

ksundong reviewed (approved):
Roach 고생하셨습니다!! 코멘트 몇 개 작성해서 의견드립니다.
다음 단계 진행해주시면 됩니다!

하지만 커밋 제목이 많이 아쉽네요.
[이 글](https://meetup.toast.com/posts/106) 한 번 읽어보시고, 더 좋은 커밋 제목을 한 번 고민해주시면 좋을 것 같습니다.

고생하셨습니다! Roach도 명절 잘 보내세요!
